### PR TITLE
feat(rpc): implementing staking reads and replace sdk.Address

### LINKED ISSUE #986: Implement staking module reads
https://github.com/cosmos/cosmos-sdk/blob/71035879e4af4551b9bcf16f79bc52785793602e/x/staking/types/querier.go#L17-L18

and

https://github.com/cosmos/cosmos-sdk/blob/71035879e4af4551b9bcf16f79bc52785793602e/x/staking/types/querier.go#L13

### PR DESCRIPTION
**Pull request by distractedm1nd**
Closes #986.

Also removes the abstraction of `sdk.Address` to make it explicit that our queries only support their corresponding address type, `AccAddress` or `ValAddress`

### CODE REVIEW

File: service/rpc/state.go [hunk 1]
```diff
@@ -348,3 +356,98 @@ func (h *Handler) handleRedelegation(w http.ResponseWriter, r *http.Request) {
...
+	vars := mux.Vars(r)
+	addrStr, exists := vars[addrKey]
+	if !exists {
+		writeError(w, http.StatusBadRequest, queryUnbondingEndpoint, errors.New("address not specified"))
+		return
+	}
+
+	// convert address to Address type
+	addr, addrerr := types.ValAddressFromBech32(addrStr)
+	if addrerr != nil {
```
Comment (renaynay) [inline service/rpc/state.go:402]: same here

File: service/state/core_access.go [hunk 2]
```diff
@@ -332,3 +336,58 @@ func (ca *CoreAccessor) Delegate(
...
+		DelegatorAddr: delAddr.String(),
+		ValidatorAddr: valAddr.String(),
+	})
+}
+func (ca *CoreAccessor) QueryRedelegations(
+	ctx context.Context,
+	srcValAddr,
+	dstValAddr Address,
+) (*stakingtypes.QueryRedelegationsResponse, error) {
+	_, ok := srcValAddr.(sdktypes.ValAddress)
```
Comment (renaynay) [inline service/state/core_access.go:380]: same here

File: service/rpc/state.go [hunk 3]
```diff
@@ -75,6 +78,11 @@ type cancelUnbondRequest struct {
 	GasLimit uint64 `json:"gas_limit"`
 }
 
+type queryRedelegationRequest struct {
```
Comment (renaynay) [inline service/rpc/state.go:81]: `queryRedelegationsRequest`
```suggestion
type queryRedelegationsRequest struct {
```

**Review by walldiss — DISMISSED** [review]

File: service/state/core_access.go [hunk 4]
```diff
@@ -332,3 +336,58 @@ func (ca *CoreAccessor) Delegate(
 	}
 	return ca.SubmitTx(ctx, signedTx)
 }
+
+func (ca *CoreAccessor) QueryDelegation(
+	ctx context.Context,
+	valAddr Address,
+) (*stakingtypes.QueryDelegationResponse, error) {
+	_, ok := valAddr.(sdktypes.ValAddress)
```
[the remark this answers is not in the document]
Comment (distractedm1nd) [reply service/state/core_access.go:344]: Well, we still cast the `sdk.Address` to `AccAddress` for things like balances, this is just an extra sanity check. I would say if we change the parameter to ValAddress, we should change them in the other methods to AccAddress as well, no?

File: service/rpc/state.go (hunk 3 above)
Comment (distractedm1nd) [reply service/rpc/state.go]: Also renamed `queryRedelegationEndpoint` to `queryRedelegationsEndpoint`

File: service/state/core_access.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (renaynay) [reply service/state/core_access.go:344]: Could do for consistency, sure.

[the remark this answers is not in the document]
Comment (distractedm1nd) [reply service/state/core_access.go]: done

File: service/state/state.go [hunk 5]
```diff
@@ -15,8 +15,11 @@ type Tx = coretypes.Tx
 // TxResponse is an alias to the TxResponse type from Cosmos-SDK.
 type TxResponse = sdk.TxResponse
 
-// Address is an alias to the Address type from Cosmos-SDK.
-type Address = sdk.Address
+// ValAddress is an alias to the ValAddress type from Cosmos-SDK.
+type ValAddress = sdk.ValAddress
+
+// AccAddress is an alias to the AccAddress type from Cosmos-SDK.
+type AccAddress = sdk.AccAddress
```
Comment (Wondertan) [inline service/state/state.go:22]: What's the reason for this change?

**Review by Wondertan — COMMENTED** [review]
The RPC part is LGTM to me. I am a bit concerned about types change here

File: service/state/state.go (hunk 5 above)
Comment (distractedm1nd) [reply service/state/state.go:22]: CC: @renaynay 

To make it explicit what account type is accepted for each request, instead of being hidden in the implementation

Comment (Wondertan) [reply service/state/state.go:22]: Feel conflicted about this refactor slipping through. I like explicitness in the case where we work with ValAddress in delegation related methods. However, the Address abstraction should still be preserved for Transfer and Balance methods to allows operations for any kind of addresses!

Comment (distractedm1nd) [reply service/state/state.go:22]: it doesn't though. transactions and balances were already only using AccAddresses before

Comment (distractedm1nd) [reply service/state/state.go:22]: (they do AccAddressFromBech32 which only accepts user accounts)

Comment (Wondertan) [reply service/state/state.go:22]: We doesn't, but it does not meant that we shouldn't or won't. Current abstraction allows this, this _uninvited_ refactor prevents this. Once there is a user requesting feature to get balance of validator, we will readd this. 

If we really want to make this PR complete all features:
Extending support for ValAddress via Address abstraction > refactoring the code to remove abstraction at all.

Comment (distractedm1nd) [reply service/state/state.go:22]: Sorry but this was invited, it was part of rene's review. Let's discuss in Sync today.

Comment (renaynay) [reply service/state/state.go:22]: @Wondertan let me clarify: 

We have 4 things going on under the hood in state access: 
1. bank module writes (transfer, submitTx, submitPFD which is just fancy submitTx)
2. bank module reads (balance, balanceForAddress same thing)
3. staking module writes (Delegate, Redelegate, etc.)
4. staking module reads (queryDelegate, queryUnbonding, etc.)

1 and 2 access same account state tree via the bank store key (albeit in separate ways since we do proof verification for bank module reads)

3 and 4 access same staking state tree via staking store key (this happens via grpc query client)

There are different representations of addresses that are used as keys to read from the respective trees. IMO it's good to make it explicit for bank module reads vs staking module reads since the error that is returned when you try to use something other than AccAddress for bank module read from the app is quite ugly and difficult for the user to parse through (even though it's very low probability they run into it if they're using our key management utility).

Comment (renaynay) [reply service/state/state.go:22]: What we can and *should* do is clarify this in the PR title and description pls 🙏🏻 @distractedm1nd

**Review by renaynay — APPROVED** [review]
Implementation looks clean, thank you @distractedm1nd  

Let's please test every endpoint against `arabica` before we merge. 🙏🏻

File: service/state/state.go (hunk 5 above)
Comment (distractedm1nd) [reply service/state/state.go:22]: #1085

**Review by Wondertan — APPROVED** [review]

**Comment by distractedm1nd** [comment]
All endpoints including queries have been tested on arabica, BUT:
- Redelegation and Redelegation Query are not 100% tested because I can't do a redelegation with only one ValAddr. But the errors returned when redelegating to the same ValAddr are reassuring enough.
